tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/Refiner.scala

```diff
@@ -166,72 +164,101 @@ private object Refiner {
+          builder.addMethod(memberMethodsInfoCaches.getInfo(linkedMethod))
+        for (linkedMethod <- linkedClass.abstractMethods)
+          builder.addMethod(abstractMethodsInfoCaches.getInfo(linkedMethod))
+        for (linkedMember <- linkedClass.exportedMembers)
+          builder.addMethod(exportedMembersInfoCaches.getInfo(linkedMember))
+
+        if (linkedClass.topLevelExports.nonEmpty) {
+          /* We do not cache top-level exports, because they're quite rare,
+           * and usually quite small when they exist.
+           */
+          builder.setIsExported(true)
+
+          val optInfo = Infos.generateTopLevelExportsInfo(
+              linkedClass.encodedName, linkedClass.topLevelExports.map(_.value))
+          optInfo.foreach(builder.addMethod(_))
         }
+
+        info = builder.result()
       }
     }
 
-    private def isUpToDate(linkedClass: LinkedClass): Boolean = {
-      versionsMatch(classVersion, linkedClass.version) &&
-      linkedMemberListVersionsMatch(staticMethodsVersions,
-          linkedClass.staticMethods) &&
-      linkedMemberListVersionsMatch(memberMethodsVersions,
-          linkedClass.memberMethods) &&
-      linkedMemberListVersionsMatch(abstractMethodsVersions,
-          linkedClass.abstractMethods) &&
-      linkedMemberListVersionsMatch(exportedMembersVersions,
-          linkedClass.exportedMembers) &&
-      linkedTopLevelExportListVersionsMatch(topLevelExportsVersions,
-          linkedClass.topLevelExports)
-    }
-
-    private def linkedMemberListVersions(
-        linkedMembers: List[Versioned[MemberDef]]): List[(String, Option[String])] = {
-      linkedMembers.map(m => (m.value.encodedName, m.version))
-    }
-
-    private def linkedMemberListVersionsMatch(
-        expected: List[(String, Option[String])],
-        linkedMembers: List[Versioned[MemberDef]]): Boolean = {
-      expected.size == linkedMembers.size &&
-      expected.zip(linkedMembers).forall {
-        case (expected, linkedMember) =>
-          linkedMember.value.encodedName == expected._1 &&
-          versionsMatch(expected._2, linkedMember.version)
+    /** Returns true if the cache has been used and should be kept. */
+    def cleanAfterRun(): Boolean = {
+      val result = cacheUsed
+      cacheUsed = false
+      if (result) {
+        // No point in cleaning the inner caches if the whole class disappears
+        staticMethodsInfoCaches.cleanAfterRun()
+        memberMethodsInfoCaches.cleanAfterRun()
+        abstractMethodsInfoCaches.cleanAfterRun()
+        exportedMembersInfoCaches.cleanAfterRun()
       }
+      result
     }
+  }
 
-    private def linkedTopLevelExportListVersion(
-        topLevelExports: List[Versioned[TopLevelExportDef]]): List[(String, Option[String])] = {
-      topLevelExports.map(e => (e.value.topLevelExportName, e.version))
+  private final class LinkedMembersInfosCache(
+      val caches: mutable.Map[(Boolean, String), LinkedMemberInfoCache])
```

Comment:
So that `LinkedMembersInfosCache` can be an `AnyVal`, reducing indirections.

And yes, static and instance members can share names. We "do" IR-check this in the sense that we don't do anything about it ;) If they were exclusive, we would have to check exclusivity.